pbj-core/pbj-runtime/src/main/java/com/hedera/pbj/runtime/grpc/ServiceInterface.java

```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright (C) 2024 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hedera.pbj.runtime.grpc;
+
+import com.hedera.pbj.runtime.io.buffer.Bytes;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.util.List;
+import java.util.concurrent.Flow;
+
+/**
+ * Defines a common interface for all implementations of a gRPC {@code service}. PBJ will generate a sub-interface
+ * for each {@code service} in the protobuf schema definition files, with default implementations of each of the
+ * given methods in this interface.
+ *
+ * <p>For example, suppose I have the following protobuf file:
+ * <pre>
+ * {@code
+ * package example;
+ *
+ * service HelloService {
+ *   rpc SayHello (HelloRequest) returns (HelloResponse);
+ * }
+ *
+ * message HelloRequest {
+ *   string greeting = 1;
+ * }
+ *
+ * message HelloResponse {
+ *   string reply = 1;
+ * }
+ * }
+ * </pre>
+ *
+ * <p>From this file, PBJ will generate a {@code HelloService} interface that extends {@code ServiceInterface}:
+ * <pre>
+ * {@code
+ * public interface HelloService extends ServiceInterface {
+ *    // ...
+ *
+ *    @NonNull
+ *    HelloResponse sayHello(final @NonNull HelloRequest request);
+ *
+ *    default String serviceName() { return "HelloService"; }
+ *    default String fullName() { return "example.HelloService"; }
+ *
+ *    // ...
+ * }
+ * }
+ * </pre>
+ *
+ * In the application code, you will simply create a new class implementing the {@code HelloService} interface, and
+ * register it with your webserver in whatever way is appropriate for your webserver.
+ */
+public interface ServiceInterface {
+    interface Method {
+        String name();
+    }
+
+    interface RequestOptions {
+        String APPLICATION_GRPC = "application/grpc";
+        String APPLICATION_GRPC_PROTO = "application/grpc+proto";
+        String APPLICATION_GRPC_JSON = "application/grpc+json";
+
+        String authority();
+
+        boolean isProtobuf();
+
+        boolean isJson();
```

Comment:
Minor:
1. This is a bit limiting as more content types could become useful in the future (YAML, gzip, etc.)
2. This is error prone because an implementation of this interface could mistakenly return `true` for both the methods, and you then have to write code to handle this situation.

I think it could be better modeled with a `enum ContentType { ...; String type; }` and then have only a single method here, like `ContentType contentType()`. This enum could (now or in the future) help handle the different types, so as to hide these details from the code that actually has to parse the bytes. This would make the content type concept strongly typed, prevent potential bugs described above, and allow us to easily extend the support to more content types in the future.

Reply:
The only problem I have with an enum is that it requires a code change to support a new content type, whereas this implementation would allow a new content type to be supported without any change in the library code. Otherwise I love the suggestion. 🤔

Reply:
I'm not sure how this is possible given these binary getters above? Any code that uses them would have to be updated if/when a new type is introduced.

However, if the content type logic is fully encapsulated in the enum implementation, then no code would need changing at all, apart from the enum itself by means of adding one more constant with proper initialization arguments to implement the support of the new content type.

Reply:
Although I agree that this would complicate the compiler a bit in order to generate the enum. So depending on how important/urgent it is for us to support more content types, we may decide to do this right now or not.

However, I disagree with the point of being able to add a new content type w/o any changes to PBJ. Your current sample code would simply fall into the !isProtobuf() branch which is equal to JSON unconditionally. So it wouldn't really support a random new content type out of the box anyway, unless the compiler is updated to support it properly.